src/lib.rs

```diff
@@ -130,24 +135,49 @@ pub extern "C" fn rustls_client_session_new(
             return rustls_result::ERROR;
         }
     };
-    let client = ClientSession::new(&config, name_ref);
+    let session = ClientSession::new(&config, name_ref);
 
-    // We've succeeded. Put the client on the heap, and transfer ownership
-    // to the caller. After this point, we must return CRUSTLS_OK so the
-    // caller knows it is responsible for this memory.
-    let b = Box::new(client);
+    // We've succeeded. Put a rustls_client_session on the heap, and transfer
+    // ownership to the caller. After this point, we must return CRUSTLS_OK so
+    // the caller knows it is responsible for this memory.
+    let b = Box::new(rustls_client_session {
+        session,
+        error: "".to_string(),
+    });
     unsafe {
         *session_out = Box::into_raw(b) as *mut _;
     }
 
     return rustls_result::OK;
 }
 
+const NULL_SESSION: &str = "session was NULL";
+
+/// After a rustls_client_session method returns an error, you may call
+/// this method to get a pointer to a buffer containing a detailed error
+/// message. The contents of the error buffer will by len bytes long,
+/// UTF-8 encoded, and not NUL-terminated.
```

Comment:
NUL-terminated was my first instinct here; @alex suggested pointer+len would be better. A third way would be "caller provides the memory, and rustls writes into it." I don't have a strong intuition about the tradeoffs and/or idioms here. @alex, are you able to say more about your reasoning for pointer+len?

Reply:
I dislike perpetuating the NUL-string approach of C. It's incredibly error prone, and also tends to result in the need for extra allocations/copies.

pointer+len is far more interoperable with modern languages, and C supports it perfectly well as well.

As for caller provides memory, that's fine,  with the note that it ends with you needing to do an extra copy so if you care about error path performance, it might be unfortunate.

Reply:
The more I think about it, the more I agree with Alex --  this library isn't meant for general consumption, so being unsurprising to C programmers is much less important than safety, and so buf+ptr is a better way to go. Besides, C programmers are used to being surprised.